### PR TITLE
Implementation of Cloud-robotics singletask_learning

### LINKED ISSUE #123: Multimodal Large Model Joint Learning Algorithm: Reproduction Based on KubeEdge-Ianvs
**What would you like to be added/modified**:
A benchmark suite for multimodal large language models deployed at the edge using KubeEdge-Ianvs:
1. Modify and adapt the existing edge-cloud data collection interface to meet the requirements of multimodal data collection;
2. Implement a Multimodal Large Language Model (MLLM) benchmark suite based on Ianvs;
3. Reproduce mainstream multimodal joint learning (training and inference) algorithms and integrate them into Ianvs single-task learning;
4. (_Advanced_) Test the effectiveness of multimodal joint learning in at least one of Ianvs' advanced paradigms (lifelong learning, incremental learning, federated learning, etc.).

**Why is this needed**:
KubeEdge-Ianvs currently focuses on edge-cloud collaborative learning (training and inference) for a single modality of data. However, edge devices, such as those in autonomous vehicles, often capture multimodal data, including GPS, LIDAR, and Camera data. Single-modal learning can no longer meet the precise inference requirements of edge devices. Therefore, this project aims to integrate mainstream multimodal large model joint learning algorithms into KubeEdge-Ianvs edge-cloud collaborative learning, providing multimodal learning capabilities.

**Recommended Skills**:
TensorFlow/Pytorch, LLMs, KubeEdge-Ianvs

**Useful links**:
[KubeEdge-Ianvs](https://github.com/kubeedge/ianvs)
[KubeEdge-Ianvs Benchmark Test Cases](https://github.com/kubeedge/ianvs/blob/main/examples/robot-cityscapes-synthia/lifelong_learning_bench/semantic-segmentation/README.md)
[Building Edge-Cloud Synergy Simulation Environment with KubeEdge-Ianvs](https://github.com/kubeedge/ianvs/blob/main/docs/guides/how-to-build-simulation-env.md)
[Artificial Intelligence - Pretrained Models Part 2: Evaluation Metrics and Methods](https://std.samr.gov.cn/gb/search/gbDetailed?id=0DF2C51A80293207E06397BE0A0AF1DA)
[Example LLMs Benchmark List](https://github.com/terryyz/llm-benchmark)
[awesome-multimodal-ml](https://github.com/pliang279/awesome-multimodal-ml)
[Awesome-Multimodal-Large-Language-Models](https://github.com/BradyFU/Awesome-Multimodal-Large-Language-Models)

### PR DESCRIPTION
**Pull request by aryan0931**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
Implementation of Cloud-robotics singletask_learning

**Which issue(s) this PR fixes**:
Fixes #123

### CODE REVIEW

**Review by MooreZheng — COMMENTED** [review]
Good to see the implementation~ @aryan0931 is suggested to introduce the implementation for KubeEdge Reviewers on the routine meeting and @hsj576 might want to take a deeper look at the implementation

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+        if type(data) is np.ndarray:
+            data = data.tolist()
+
+        self.validator.test_loader = DataLoader(
+            data, batch_size=self.validator.args.test_batch_size, shuffle=False, pin_memory=True)
+        return self.validator.validate()
+
+    def load(self, model_url):
+        model_url = '/home/wxc/dev/ianvs/models/model_best_mapi_only.pth'
```
Comment (MooreZheng) [inline examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:149]: The model_url is hardcoded in the load function which would bypass the parameter, which should be revised.

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                # save checkpoint when it meets eval_interval or the training finished
+                is_best = False
+                self.checkpoint_path = self.trainer.saver.save_checkpoint({
+                    'epoch': epoch + 1,
+                    'state_dict': self.trainer.model.state_dict(),
+                    'optimizer': self.trainer.optimizer.state_dict(),
+                    'best_pred': self.trainer.best_pred,
+                }, is_best)
+
+            # if not self.trainer.args.no_val and \
```
Comment (MooreZheng) [inline examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:102]: Considering Clean Code: remove redundant codes instead of making them as comment

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            raise Exception("Train data is None.")
+
+        self.trainer.train_loader, self.trainer.val_loader, _, _ = \
+            make_data_loader(
+                self.trainer.args, train_data=train_data, valid_data=valid_data, **kwargs)
+
+        # Define lr scheduler
+        self.trainer.scheduler = LR_Scheduler(self.trainer.args.lr_scheduler, self.trainer.args.lr,
+                                              self.trainer.args.epochs, len(self.trainer.train_loader))
+        print("Total epoches:", self.trainer.args.epochs)
```
Comment (MooreZheng) [inline examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:83]: should replace print using loggers

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testenv/accuracy.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        # Add batch sample into evaluator
+        evaluator.add_batch(target, y_pred[i])
+
+    # Test during the training
+    # Acc = evaluator.Pixel_Accuracy()
+    CPA = evaluator.Pixel_Accuracy_Class()
+    mIoU = evaluator.Mean_Intersection_over_Union()
+    FWIoU = evaluator.Frequency_Weighted_Intersection_over_Union()
+
+    print("CPA:{}, mIoU:{}, fwIoU: {}".format(CPA, mIoU, FWIoU))
```
Comment (MooreZheng) [inline examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testenv/accuracy.py:57]: should replace print using loggers

**Review by MooreZheng — CHANGES_REQUESTED** [review]
It is a basic version of the multi-modal model project. Before merging the PR, 
1) as mentioned at the routine meeting, an output demonstration is needed. 
2) the above comments should be fixed.

**Comment by MooreZheng** [comment]
Another new LFX term is coming. Any progress for this PR? @aryan0931

**Comment by aryan0931** [comment]
> Another new LFX term is coming. Any progress for this PR? @aryan0931 

Sorry sir I was not well from some days, I will raise the final Pull request till Sunday.

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py (hunk 1 above)
Comment (MooreZheng) [reply examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:149]: Not solved yet

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py (hunk 2 above)
Comment (MooreZheng) [reply examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:102]: Not solved yet

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py (hunk 3 above)
Comment (MooreZheng) [reply examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:83]: Not fully solved yet. There are still "print" functions in the code.

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
+# Copyright 2022 The KubeEdge Authors.
```
[the remark this answers is not in the document]
Comment (MooreZheng) [reply examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testalgorithms/basemofel.py:1]: Not resolved yet. The file name is still "basemofel".

File: examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testenv/accuracy.py (hunk 4 above)
Comment (MooreZheng) [reply examples/Cloud_Robotics/singletask_learning_bench/Semantic_Segmentation/testenv/accuracy.py]: The solution of metrics of the new version from PR #176 looks fine to me
